File: templates/problems/03_sliding_window/05_min_window_substring/min_window_substring_test.go
package min_window_substring

import "testing"

func TestMinWindow(t *testing.T) {
	tests := []struct {
		name     string
		s        string
		t_str    string
		expected string
	}{
		{
			name:     "basic case",
			s:        "ADOBECODEBANC",
			t_str:    "ABC",
			expected: "BANC",
		},
		{
			name:     "exact match",
			s:        "a",
			t_str:    "a",
			expected: "a",
		},
		{
			name:     "no valid window",
			s:        "a",
			t_str:    "aa",
			expected: "",
		},
		{
			name:     "t not in s",
			s:        "abc",
			t_str:    "z",
			expected: "",
		},
		{
			name:     "entire string is window",
			s:        "abc",
			t_str:    "abc",
			expected: "abc",
		},
		{
			name:     "duplicate chars in t",
			s:        "aaabbc",
			t_str:    "aab",
			expected: "aab",
		},
		{
			name:     "empty s",
			s:        "",
			t_str:    "a",
			expected: "",
		},
	}
	for _, tt := range tests {
		t.Run(tt.name, func(t *testing.T) {
			got := minWindow(tt.s, tt.t_str)
			if got != tt.expected {
				t.Errorf("minWindow(%q, %q) = %q, want %q", tt.s, tt.t_str, got, tt.expected)
			}
		})
	}
}

File: templates/problems/03_sliding_window/05_min_window_substring/typescript/minWindowSubstring.test.ts
import { describe, it, expect } from 'vitest';
import { minWindow } from './minWindowSubstring';

describe('MinWindowSubstring', () => {
  it('basic case', () => {
    expect(minWindow('ADOBECODEBANC', 'ABC')).toBe('BANC');
  });

  it('exact match', () => {
    expect(minWindow('a', 'a')).toBe('a');
  });

  it('no valid window', () => {
    expect(minWindow('a', 'aa')).toBe('');
  });

  it('t not in s', () => {
    expect(minWindow('abc', 'z')).toBe('');
  });

  it('entire string is window', () => {
    expect(minWindow('abc', 'abc')).toBe('abc');
  });

  it('duplicate chars in t', () => {
    expect(minWindow('aaabbc', 'aab')).toBe('aab');
  });

  it('empty s', () => {
    expect(minWindow('', 'a')).toBe('');
  });

  it('many duplicates before target chars', () => {
    expect(minWindow('aaaaaaaaaaaabbbbbcdd', 'abcdd')).toBe('abbbbbcdd');
  });
});
